stream: need to cleanup event listeners if last stream is readable

meixg opened the pull request:
fix: https://github.com/nodejs/node/issues/35452

```javascript
const {pipeline, Duplex, PassThrough, Writable} = require('stream');

const a = new PassThrough();
a.end('foobar');

const b = new Duplex({
    write(chunk, encoding, callback) {
        callback();
    }
});

pipeline(a, b, error => {
    if (error) {
        throw error;
    }
    console.log(b.listenerCount('error'));
    setTimeout(() => {
        console.log(b.listenerCount('error'));
        b.destroy(new Error('no way'));
    }, 100);
});
```

output:

```
0
0
node:events:505
      throw er; // Unhandled 'error' event
      ^

Error: no way
```

@ronag @mcollina

ronag reviewed (changes requested):
This is subtle problems. I think a possible solution is to just cleanup the listeners for the last streams.

meixg commented:
> This has subtle problems. I think a possible solution is to just cleanup the listeners for the last stream.

This change uses an `isLastStrem` variable to only process the last stream.

meixg commented:
> Can you please add a unit test?

Unit test added :)

ronag reviewed (changes requested):
You need to check if the stream is readable at the end before the cleanup.

ronag commented on lib/internal/streams/pipeline.js:
```diff
@@ -45,14 +45,17 @@ function destroyer(stream, reading, writing) {
-    finished = true;
-    destroyImpl.destroyer(stream, err || new ERR_STREAM_DESTROYED('pipe'));
+  return {
+    callback: (err) => {
```
```suggestion
    destroy: (err) => {
```

ronag commented on lib/internal/streams/pipeline.js:
```diff
@@ -194,6 +201,11 @@ function pipelineImpl(streams, callback, opts) {
 
     if (final) {
+      if (!error) {
+        while (lastStreamCleanup && lastStreamCleanup.length) {
+          lastStreamCleanup.shift()();
+        }
```
```suggestion
        lastStreamCleanup?.forEach(fn => fn());
```

ronag commented on lib/internal/streams/pipeline.js:
```diff
@@ -204,22 +216,34 @@ function pipelineImpl(streams, callback, opts) {
     if (isNodeStream(stream)) {
       if (end) {
-        destroys.push(destroyer(stream, reading, writing));
+        const { callback, cleanup } = destroyer(stream, reading, writing);
+        destroys.push(callback);
+
+        if (lastStreamCleanup && isLastStream) {
+          lastStreamCleanup.push(cleanup);
+        }
```
```suggestion
        const { destroy, cleanup } = destroyer(stream, reading, writing);
        destroys.push(destroy);

        if (isLastStream) {
          lastStreamCleanup?.push(cleanup);
        }
```

ronag commented on lib/internal/streams/pipeline.js:
```diff
@@ -204,22 +216,34 @@ function pipelineImpl(streams, callback, opts) {
-      });
+      }
+      stream.on('error', onError);
+      if (lastStreamCleanup && isLastStream) {
+        lastStreamCleanup.push(() => {
+          stream.removeListener('error', onError);
+        });
+      }
```
```suggestion
      if (isLastStream) {
        lastStreamCleanup?.push(() => {
          stream.removeListener('error', onError);
        });
      }
```

ronag commented on lib/internal/streams/pipeline.js:
```diff
@@ -285,12 +309,15 @@ function pipelineImpl(streams, callback, opts) {
         ret = pt;
 
-        destroys.push(destroyer(ret, false, true));
+        destroys.push(destroyer(ret, false, true).callback);
```
```suggestion
        const { destroy, cleanup } = (destroyer(ret, false, true)
        destroys.push(destroy);
        if (isLastStream) {
          lastStreamCleanup?.push(cleanup);
        }
```

ronag commented on lib/internal/streams/pipeline.js:
```diff
@@ -285,12 +309,15 @@ function pipelineImpl(streams, callback, opts) {
         finishCount += 2;
-        pipe(ret, stream, finish, { end });
+        const cleanup = pipe(ret, stream, finish, { end });
+        if (lastStreamCleanup && isLastStream) {
+          lastStreamCleanup.push(cleanup);
+        }
```
```suggestion
        if (isLastStream) {
          lastStreamCleanup?.push(cleanup);
        }
```

ronag reviewed (commented):
Maybe ok. A little scary. Added some fixes.

ronag commented on lib/internal/streams/pipeline.js:
```diff
@@ -159,6 +162,10 @@ function pipelineImpl(streams, callback, opts) {
 
+  // Need to cleanup event listeners if last stream is readable
+  // https://github.com/nodejs/node/issues/35452
+  const lastStreamCleanup = streams[streams.length - 1].readable && [];
```
```suggestion
  const lastStreamCleanup = isReadableNodeStream(streams[streams.length - 1]) && [];
```

ronag reviewed (changes requested):
Will also need a doc update.

meixg commented:
> Maybe ok. A little scary. Added some fixes.

Applied suggestions from code review, and added a test case for async generators.

The doc is updated too.

mcollina reviewed (approved):
lgtm

meixg commented:
May need a rerun of the ci.

meixg commented:
Is the ci broken?

meixg commented:
I checked that the CI's failure is not caused by the change. Maybe rerun again...?

meixg commented:
Can you help me with this CI? @ronag

meixg commented:
only one case failing now, pls help me with this... @ronag 
![image](https://user-images.githubusercontent.com/9262426/154012965-4af35a2e-67dd-4604-bd4c-7a62ef139764.png)

ronag commented:
@Trott is that flakiness?

meixg commented:
CI passed🎉

meixg commented:
> This PR needs to wait 25 more minutes to land

can land now :)

meixg commented:
May need to use "commit-queue-squash"

meixg commented:
Seems that the "commit-queue-squash" needs to work with "commit-queue"

meixg commented:
thanks! @ronag
